### PR TITLE
Keyboard shortcut editor enhancements

### LINKED ISSUE #2475: Allow sorting by column of Keyboard Shortcuts.
It would be convenient if we could sort the Keyboard Shortcuts by column.

It would make it a lot easier to see if a given key is already used. For example, I could see all the items that are bound to F12 just by scrolling down to F12.

It was pretty easy to add, only having to change the plugins/shortcut.py module. I'll submit a PR soon, once I figure out how to do it... :-P

### PR DESCRIPTION
**Pull request by goanpeca**
Fix #2475 
Fix some pending issues from https://github.com/spyder-ide/spyder/pull/2367

----
## "Fuzzy" string matching

image

image

image

## Todo
- [x] Fix docstrings
- [x] Change toggle behavior in shortcut dialog
- [x] Add fuzzy sorting
- [x] Change information display for conflicting shortcuts
- [x] Handle sequences bigger than 4 keys
- [x] Relocate HTMLDelegate to `widgets/helperwidgets.py`

## For another PR (when I have access to a Mac...)
- Fix text color on selected text (in Mac)
- Fix dead strokes input

### CODE REVIEW

**Comment by goanpeca** [comment]
@SylvainCorlay can you give me an example of what happens with the dead stroke issues?

What type of keyboard layout you have in your pc... OS.. etc..

Cheers

**Comment by ccordoba12** [comment]
@goanpeca: One example that @SylvainCorlay showed me in person: pressing `Ctrl + <tilde> + e` where `<tilde>` is the key to get `é` on the screen. That's a dead key :-)

**Comment by SylvainCorlay** [comment]
@goanpeca if you use a `US-international` keyboard layout, the following key sequences do

`` ` `` then `e`   -> `è`
`` ^ `` then `e`   -> `ê`
`` ' `` then `e`   -> `é`

meaning that the first character is a dead key and does not do anything by itself. If you just want a quote, you need to press the space bar.

If you use a Spanish layout, I think that `^` is already a dead key.

I don't think that we really need to support dead keys for keyboard shortcuts. Although in the French layout, things like `é` are a single key stroke, and should not create any problem.

**Comment by goanpeca** [comment]
I cannot reproduce that behavior.

On ubuntu with US-International, if I press Control+' and (while holding control) press "e", I get
`Ctrl+'; Ctrl+e`

**Comment by ccordoba12** [comment]
@goanpeca, do you have a Latin American keyboard to test? This is what I'm getting with `Ctrl+<tilde>+e`:

    Ctrl+ោ

**Comment by goanpeca** [comment]
Not right now...

**Comment by ccordoba12** [comment]
Ok, no prob. We can check it out week when we meet in Colombia :-)

**Comment by goanpeca** [comment]
Any comments on the search order screenshots... ?

File: spyderlib/plugins/shortcuts.py [hunk 1]
```diff
@@ -15,30 +15,29 @@
...
 
 # Local imports
 from spyderlib.baseconfig import _, debug_print
 from spyderlib.guiconfig import (get_shortcut, set_shortcut,
                                  iter_shortcuts, reset_shortcuts)
 from spyderlib.plugins.configdialog import GeneralConfigPage
+from spyderlib.utils import icon_manager as ima
 from spyderlib.utils.qthelpers import get_std_icon
+from spyderlib.utils.stringmatching import get_search_regex, get_search_scores
 from spyderlib.widgets.arraybuilder import HelperToolButton
```
Comment (ccordoba12) [inline spyderlib/plugins/shortcuts.py]: This helper button should be moved to `helperwidgets` too. But this is just a reminder, let's leave it for another PR :-)

File: spyderlib/plugins/shortcuts.py [hunk 2]
```diff
@@ -74,57 +75,13 @@ def keyReleaseEvent(self, e):
...
-        options = QStyleOptionViewItem(option)
-        self.initStyleOption(options, index)
-
-        doc = QTextDocument()
-        doc.setHtml(options.text)
-        doc.setTextWidth(options.rect.width())
-
-        return QSize(doc.idealWidth(), doc.size().height())
-
-
```
Comment (ccordoba12) [inline spyderlib/plugins/shortcuts.py]: Nice to see this moved out of here ;-)

File: spyderlib/plugins/shortcuts.py [hunk 3]
```diff
@@ -151,10 +108,15 @@ def keyPressEvent(self, event):
             super(ShortcutFinder, self).keyPressEvent(event)
 
 
+NO_WARNING, SEQUENCE_LENGTH, SEQUENCE_CONFLICT, INVALID_KEY = [0, 1, 2, 3]
```
Comment (ccordoba12) [inline spyderlib/plugins/shortcuts.py]: Error codes? Please add a comment above this line to explain their meaning

File: spyderlib/plugins/shortcuts.py [hunk 4]
```diff
@@ -289,12 +255,20 @@ def nonedit_keyrelease(self, e):
...
         self.npressed -= 1
         if self.npressed <= 0:
-            if len(self.keys) == 1 and (list(self.keys)[0] == Qt.Key_Tab):
+            key = e.key()
+
+            if len(self.keys) == 1 and key == Qt.Key_Tab:
                 self.toggle_state()
 
+            if len(self.keys) == 1 and key == Qt.Key_Escape:
+                self.close()
```
Comment (ccordoba12) [inline spyderlib/plugins/shortcuts.py]: Instead of closing the dialog, my suggestion for `Esc` is to do `toggle_state` (as `Tab`) **and** also reset the shortcut present in the new shortcut QLineEdit

File: spyderlib/plugins/shortcuts.py [hunk 5]
```diff
@@ -318,43 +292,65 @@ def check_conflicts(self):
...
             for s in conflicts:
-                tip_body += ' - {0}: {1}\n'.format(s.context, s.name)
-            tip = '{0}{1}'.format(tip_title, tip_body)
+                tip_body += ' - {0}: {1}<br>'.format(s.context, s.name)
+            tip = template.format(tip_title, tip_body)
+            warn = True
+        elif warning_type == SEQUENCE_LENGTH:
+            template = '<i>{0}</i>'
+            tip = _('A compound sequence can have {break} a maximum of '
+                    '4 subsequences.{break}').format(**{'break': '<br>'})
```
Comment (ccordoba12) [inline spyderlib/plugins/shortcuts.py]: When will this situation be encountered? Could you give an example?

File: spyderlib/plugins/shortcuts.py [hunk 6]
```diff
@@ -176,6 +138,7 @@ def __init__(self, parent, context, name, sequence, shortcuts):
         self.text_new_sequence = CustomLineEdit(self)
         self.text_new_sequence.setPlaceholderText(sequence)
         self.helper_button = HelperToolButton()
+        self.label_warning = QLabel('<br><br>')
```
Comment (ccordoba12) [inline spyderlib/plugins/shortcuts.py]: I think this widget should be initialized just as `QLabel()` and hidden by default, unless there's a warning to show.

I say it because right now the dialog looks like this

seleccion_003

which (to me) it's a bit ugly because there's too much space between the line edit and the buttons below it

File: spyderlib/plugins/shortcuts.py [hunk 7]
```diff
@@ -415,14 +417,17 @@ def save(self):
         set_shortcut(self.context, self.name, self.key)
 
 
-CONTEXT, NAME, SEQUENCE = [0, 1, 2]
+CONTEXT, NAME, SEQUENCE, SEARCH_SCORE = [0, 1, 2, 3]
```
Comment (ccordoba12) [inline spyderlib/plugins/shortcuts.py]: I'm getting this message when trying to filter shortcuts:

```python-traceback
Traceback (most recent call last):
  File "/home/carlos/Projects/spyder/github-repo/spyderlib/plugins/shortcuts.py", line 698, in set_regex
    self.sortByColumn(SEARCH_SCORE)
TypeError: QTableView.sortByColumn(int, Qt.SortOrder): not enough arguments
```

and I think that's because we have three columns in the shortcuts table, and so SEARCH_NAME is not a valid column number.

File: spyderlib/plugins/shortcuts.py [hunk 8]
```diff
@@ -487,20 +470,24 @@ def data(self, index, role=Qt.DisplayRole):
...
                     else:
                         color = self.text_color
-                return to_qvariant(self._enrich_text(shortcut.name, color))
+                text = self.rich_text[row]
+                text = '<p style="color:{0}">{1}</p>'.format(color, text)
+                return to_qvariant(text)
             elif column == SEQUENCE:
                 text = QKeySequence(key).toString(QKeySequence.NativeText)
                 return to_qvariant(text)
+            elif column == SEARCH_SCORE:
```
Comment (ccordoba12) [inline spyderlib/plugins/shortcuts.py]: I don't understand why you're treating search scores as a new column :-)

Comment (ccordoba12) [inline spyderlib/plugins/shortcuts.py]: It looks like a hack...

Comment (goanpeca) [inline spyderlib/plugins/shortcuts.py]: Well call it what you want ;-).... but the idea is that to filter what is displayed in the `QTableView` I use a `SortFilterProxy` subclass which filters (based on a regex in this case), this handles the rows that get displayed... now for the order...

...having a column that stores the score makes the code simpler cause I can rely on the builtin sort that comes by default in the `SortFilterProxy`.

File: spyderlib/plugins/shortcuts.py (hunk 7 above)
Comment (goanpeca) [inline spyderlib/plugins/shortcuts.py]: I guess you are using Qt5 ?, and the order of the sort is a mandatory argument there, simple fix,

File: spyderlib/plugins/shortcuts.py (hunk 6 above)
Comment (goanpeca) [inline spyderlib/plugins/shortcuts.py]: Ok I will add only one break, otherwise the dialog will resize if any error is displayed below....

Comment (goanpeca) [inline spyderlib/plugins/shortcuts.py]: and resizing on the go is even uglier

File: spyderlib/plugins/shortcuts.py (hunk 5 above)
Comment (goanpeca) [inline spyderlib/plugins/shortcuts.py]: Type `Ctrl+1, Ctrl+2, Ctrl+3, Ctrl+4, Ctrl+5` (all while holding control and in quick succession) This shortcut has more than for and is not accepted, hence it will display and error.

File: spyderlib/plugins/shortcuts.py (hunk 6 above)
Comment (ccordoba12) [inline spyderlib/plugins/shortcuts.py]: @goanpeca, please move forward with this (and my other suggestions) to have this ready for beta1 :-)

File: spyderlib/plugins/shortcuts.py (hunk 7 above)
Comment (ccordoba12) [inline spyderlib/plugins/shortcuts.py]: Yep, using Qt5 here ;-)

File: spyderlib/plugins/shortcuts.py (hunk 8 above)
Comment (ccordoba12) [inline spyderlib/plugins/shortcuts.py]: Ok, I see. Please add a comment to the code explaining this.

**Comment by ccordoba12** [comment]
@goanpeca, let's leave the *dead keystroke* task for another PR.

@blink1073, would you mind to take a look at the fuzzy matching code introduced by @goanpeca in `utils/stringmatching.py`?

File: spyderlib/utils/stringmatching.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+
+def get_search_score(query, choice, ignore_case=True, apply_regex=True,
+                     template='{0}'):
+    """ Returns a tuples with the enriched text (if a template is provided)
+    and a score for the match.
+
+    Parameters
+    ----------
+    query : str
+        String with letters to seacrh in choice (in order of appearance).
```
Comment (blink1073) [inline spyderlib/utils/stringmatching.py]: search*

File: spyderlib/utils/stringmatching.py [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+    choice : string 
+        Sentences/words in which to search for the 'query' letters.
+    ignore_case : True
+        Optional value perform a case insensitive search (True by default).
+    apply_regex : True
+        Optional value (True by default) to perform a regex search. Usefull
```
Comment (blink1073) [inline spyderlib/utils/stringmatching.py]: Useful*

File: spyderlib/utils/stringmatching.py [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+        when this function is called directly.
+    template : str
+        Optional template string to surround letters found in choices. This is
+        useful when using a rich text editor ('{}' by default).
+        Examples: '<b>{}</b>', '<code>{}</code>', '<i>{}</i>'
+
+    Returns
+    -------
+    results : tupple
```
Comment (blink1073) [inline spyderlib/utils/stringmatching.py]: tuple*

File: spyderlib/utils/stringmatching.py [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+                      valid_only=False, sort=False):
+    """Search for query inside choices and return a list of tuples.
+    
+    Returns a list of tuples of text with the enriched text (if a template is
+    provided) and a score for the match. Lower scores imply a better match.
+
+    Parameters
+    ----------
+    query : str
+        String with letters to seacrh in each choice (in order of appearance).
```
Comment (blink1073) [inline spyderlib/utils/stringmatching.py]: search*

File: spyderlib/utils/stringmatching.py [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+    ignore_case : True
+        Optional value perform a case insensitive search (True by default).
+    template : str
+        Optional template string to surround letters found in choices. This is
+        useful when using a rich text editor ('{}' by default).
+        Examples: '<b>{}</b>', '<code>{}</code>', '<i>{}</i>'
+
+    Returns
+    -------
+    results : list of tupples
```
Comment (blink1073) [inline spyderlib/utils/stringmatching.py]: tuples*

File: spyderlib/utils/stringmatching.py [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+        pattern = re.compile(regex, re.IGNORECASE)
+    else:
+        pattern = re.compile(regex)
+
+    return pattern
+
+
+def get_search_score(query, choice, ignore_case=True, apply_regex=True,
+                     template='{0}'):
+    """ Returns a tuples with the enriched text (if a template is provided)
```
Comment (blink1073) [inline spyderlib/utils/stringmatching.py]: tuple*

File: spyderlib/utils/stringmatching.py [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+    and a score for the match.
+
+    Parameters
+    ----------
+    query : str
+        String with letters to seacrh in choice (in order of appearance).
+    choice : string 
+        Sentences/words in which to search for the 'query' letters.
```
Comment (blink1073) [inline spyderlib/utils/stringmatching.py]: Sentence*

File: spyderlib/utils/stringmatching.py [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+    ignore_case : True
```
Comment (blink1073) [inline spyderlib/utils/stringmatching.py]: ignore_case : boolean, optional
       Perform a case insensitive search (True by default).

**Comment by ccordoba12** [comment]
@goanpeca, please address @blink1073 and my comments so that we can merge this for beta1. I think we need to merge this before releasing it because it has important improvements.

**Comment by goanpeca** [comment]
@ccordoba12 I made the corrections, can you test on your side?

Cheers

**Comment by ccordoba12** [comment]
@goanpeca, I have a patch (against your latest commit in this branch) that solves all issues I have right now with the `ShortcutEditor` dialog:

```patch
@@ -139,7 +139,8 @@ class ShortcutEditor(QDialog):
         self.text_new_sequence = CustomLineEdit(self)
         self.text_new_sequence.setPlaceholderText(sequence)
         self.helper_button = HelperToolButton()
-        self.label_warning = QLabel('<br>')
+        self.label_warning = QLabel()
+        self.label_warning.hide()
 
         bbox = QDialogButtonBox(QDialogButtonBox.Ok | QDialogButtonBox.Cancel)
         self.button_ok = bbox.button(QDialogButtonBox.Ok)
@@ -164,7 +165,7 @@ class ShortcutEditor(QDialog):
         self.label_warning.setFocusPolicy(Qt.NoFocus)
 
         # Layout
-        spacing = 24
+        spacing = 5
         layout_sequence = QGridLayout()
         layout_sequence.addWidget(self.label_info, 0, 0, 1, 3)
         layout_sequence.addItem(QSpacerItem(spacing, spacing), 1, 0, 1, 2)
@@ -267,8 +268,8 @@ class ShortcutEditor(QDialog):
 
             if len(self.keys) == 1 and key == Qt.Key_Escape:
                 self.set_sequence('')
-                self.toggle_state()
-                return
+                self.label_warning.setText(_("Please introduce a different "
+                                             "shortcut"))
 
             if len(self.keys) == 1 and key in [Qt.Key_Return, Qt.Key_Enter]:
                 self.toggle_state()
@@ -278,7 +279,7 @@ class ShortcutEditor(QDialog):
                 self.nonedit_keyrelease(e)
             else:
                 debug_print('keys: {}'.format(self.keys))
-                if self.keys:
+                if self.keys and key != Qt.Key_Escape:
                     self.validate_sequence()
                 self.keys = set()
                 self.key_modifiers = set()
@@ -304,13 +305,14 @@ class ShortcutEditor(QDialog):
 
         if warning_type == NO_WARNING:
             warn = False
-            tip = '<br><br>'
+            tip = ''
         elif warning_type == SEQUENCE_CONFLICT:
             template = '<i>{0}<b>{1}</b></i>'
             tip_title = _('The new shorcut conflicts with:') + '<br>'
             tip_body = ''
             for s in conflicts:
                 tip_body += ' - {0}: {1}<br>'.format(s.context, s.name)
+            tip_body = tip_body[:-4]  # Removing last <br>
             tip = template.format(tip_title, tip_body)
             warn = True
         elif warning_type == SEQUENCE_LENGTH:
@@ -326,6 +328,7 @@ class ShortcutEditor(QDialog):
             warn = True
 
         if warn:
+            widget_message.show()
             widget.setIcon(get_std_icon('MessageBoxWarning'))
         else:
             widget.setIcon(get_std_icon('DialogApplyButton'))
```

This patch:
1. Fixes the spacing issues in the dialog (i.e. a huge empty space between the shortcut line edit and the buttons).
2. Uses `Esc` to reset a conflicting shortcut, and let the user introduce a new one at the same time.

I'm fine with hidding `label_warning` at first and then showing it when a conflict appears.

**Comment by ccordoba12** [comment]
Forget about my last patch, I have other improvements so I'll do a PR against yours :-)

**Comment by goanpeca** [comment]
Ok, please go ahead!

Should I make a new PR for the plugins (the pending @Nodd pr?)

**Comment by ccordoba12** [comment]
Yes, please! ;-) We need to start moving on that front ASAP, but remember that that's work for beta2.

**Comment by goanpeca** [comment]
So what is missing for beta1 at this point?

**Comment by ccordoba12** [comment]
Take a look at our shared Google doc. There is the list of missing things.

**Comment by goanpeca** [comment]
Got it...

**Comment by goanpeca** [comment]
Ok, feel free to merge this one then :-)
